pkg/commands/loadgen.go

```diff
@@ -0,0 +1,214 @@
+		return err
+	}
+
+	writeClient, err := remote.NewClient(0, &remote.ClientConfig{
+		URL:     &config.URL{URL: writeURL},
+		Timeout: model.Duration(c.writeTimeout),
+	})
+	if err != nil {
+		return err
+	}
+	c.writeClient = writeClient
+
+	queryClient, err := api.NewClient(api.Config{
+		Address: c.queryURL,
+	})
+	if err != nil {
+		return err
+	}
+	c.queryClient = v1.NewAPI(queryClient)
+
+	http.Handle("/metrics", promhttp.Handler())
+	go http.ListenAndServe(c.metricsListenAddress, nil)
+
+	c.wg.Add(c.parallelism)
+	c.wg.Add(c.queryParallelism)
+
+	metricsPerShard := c.activeSeries / c.parallelism
+	for i := 0; i < c.activeSeries; i += metricsPerShard {
+		go c.runWriteShard(i, i+metricsPerShard)
+	}
+
+	for i := 0; i < c.queryParallelism; i++ {
+		go c.runQueryShard()
+	}
+
+	c.wg.Wait()
+	return nil
+}
+
+func (c *LoadgenCommand) runWriteShard(from, to int) {
+	defer c.wg.Done()
+	ticker := time.NewTicker(c.scrapeInterval)
+	c.runScrape(from, to)
+	for range ticker.C {
+		c.runScrape(from, to)
+	}
+}
+
+func (c *LoadgenCommand) runScrape(from, to int) {
+	for i := from; i < to; i += c.batchSize {
+		if err := c.runBatch(i, i+c.batchSize); err != nil {
+			log.Printf("error sending batch: %v", err)
+		}
+	}
+	fmt.Printf("sent %d samples\n", to-from)
+}
+
+func (c *LoadgenCommand) runBatch(from, to int) error {
+	var (
+		req = prompb.WriteRequest{
+			Timeseries: make([]prompb.TimeSeries, 0, to-from),
+		}
+		now = time.Now().UnixNano() / int64(time.Millisecond)
+	)
+
+	for i := from; i < to; i++ {
+		timeseries := prompb.TimeSeries{
+			Labels: []prompb.Label{
+				{Name: "__name__", Value: "node_cpu_seconds_total"},
+				{Name: "job", Value: "node_exporter"},
+				{Name: "instance", Value: fmt.Sprintf("instance%000d", i)},
+				{Name: "cpu", Value: "0"},
+				{Name: "mode", Value: "idle"},
+			},
+			Samples: []prompb.Sample{{
+				Timestamp: now,
+				Value:     rand.Float64(),
+			}},
+		}
+		req.Timeseries = append(req.Timeseries, timeseries)
+	}
+
+	data, err := proto.Marshal(&req)
+	if err != nil {
+		return err
+	}
+
+	compressed := snappy.Encode(nil, data)
+
+	start := time.Now()
+	if err := c.writeClient.Store(context.Background(), compressed); err != nil {
+		writeRequestDuration.WithLabelValues("error").Observe(time.Now().Sub(start).Seconds())
+		return err
+	}
+	writeRequestDuration.WithLabelValues("success").Observe(time.Now().Sub(start).Seconds())
+
+	return nil
+}
+
+func (c *LoadgenCommand) runQueryShard() {
+	defer c.wg.Done()
+	for {
+		c.runQuery()
+	}
+}
+
+func (c *LoadgenCommand) runQuery() {
+	ctx, cancel := context.WithTimeout(context.Background(), c.queryTimeout)
+	defer cancel()
+	r := v1.Range{
+		Start: time.Now().Add(-c.queryDuration),
+		End:   time.Now(),
+		Step:  time.Minute,
+	}
+	start := time.Now()
+	_, _, err := c.queryClient.QueryRange(ctx, c.query, r)
+	if err != nil {
+		queryRequestDuration.WithLabelValues("error").Observe(time.Now().Sub(start).Seconds())
+		log.Printf("error doing query: %v", err)
```

Comment:
```suggestion
		log.Printf("error doing query: %v", err)
		return
```

I think we need to return here or else the success metric will be logged as well